### PR TITLE
TMDM-14728 Hibernate & Lucene upgrade

### PR DESCRIPTION
**Pull request by hw-zhu**
https://jira.talendforge.org/browse/TMDM-14728
**What is the current behavior?** (You should also link to an open issue here)
At present, hibernate-core-4.3.8.Final.jar, hibernate-search-5.0.1.jar, lucene-core-4.10.3.jar and other some related the third package have been deprecated, so need to replace with new one. 
After upgraded, previous some API has been deleted,  MDM might doesn't work well, so need to use some recommended API.

**What is the new behavior?**
Upgrade Hibernate core to latest release version of 5.4.18.Final
Upgrade Hibernate search to latest release version of 5.11.5.Final
Upgrade Lucene to new release version of 5.5.5
In order to integrate MDM with Hibernate 5, a new Class diagram structure be designed enclosed copy. New class marked as orange to adapt new process between MDM and Hibernate 5. you can check it from jira TMDM-14728.

**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features) ?

**What kind of change does this PR introduce?**

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/mapping/MDMTableExporter.java [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    private String convertDefaultValue(Dialect dialect, String sqlType, String defaultValue) {
+        String defaultSQL = StringUtils.EMPTY;
+        if (StringUtils.isNotBlank(defaultValue) && isDefaultValueNeeded(sqlType, dialect) && !isDateType(sqlType)) {
+            defaultSQL = " DEFAULT " + defaultValue;
+        }
+        return defaultSQL;
+    }
+
+    private static boolean isDefaultValueNeeded(String sqlType, Dialect dialect) {
+        return !"longtext".equals(sqlType) || !(dialect instanceof MySQLDialect);
```
Comment (lpw82gf) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/mapping/MDMTableExporter.java:240]: Are you sure it is `MySQLDialect `, as I remember we change to use `MySQL57Dialect ` for MySQL?

Comment (hw-zhu) [reply org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/mapping/MDMTableExporter.java]: updated

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/StandardQueryHandler.java [hunk 2]
```diff
@@ -1476,6 +1492,7 @@ public Criterion visit(Compare condition) {
                 if (condition.getLeft() instanceof Field) {
                     Field leftField = (Field) condition.getLeft();
                     FieldMetadata fieldMetadata = leftField.getFieldMetadata();
+                    // TODO see how to fix this
```
Comment (htyin) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/StandardQueryHandler.java:1495]: what problem do we need fix here?

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/UpdateReportTypeMapping.java [hunk 3]
```diff
@@ -171,7 +171,11 @@ public DataRecord setValues(Wrapper from, DataRecord to) {
 
     @Override
     public String getDatabaseTimestamp() {
-        return "x_time_in_millis"; //$NON-NLS-1$
+        if (databaseUpdateReportType.getKeyFields().size() == 2) {
```
Comment (VictorLi-talendbj) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/UpdateReportTypeMapping.java:174]: Can add a comment to explain why the size == 2?

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/mapping/MDMDenormalizedTable.java [hunk 4]
```diff
@@ -57,18 +58,19 @@ public Iterator sqlAlterStrings(Dialect dialect, Mapping p, TableMetadata tableI
...
-                column.setSqlTypeCode(column.getSqlTypeCode(p));
+                column.setSqlTypeCode(column.getSqlTypeCode(metadata));
             }
             if (column.getSqlType() == null) {
-                column.setSqlType(column.getSqlType(dialect, p));
+                column.setSqlType(column.getSqlType(dialect, metadata));
             }
 
-            ColumnMetadata columnInfo = tableInfo.getColumnMetadata(column.getName());
+//            ColumnMetadata columnInfo = tableInfo.getColumnMetadata(column.getName());
```
Comment (VictorLi-talendbj) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/mapping/MDMDenormalizedTable.java:67]: Useless comment/code?

Comment (hw-zhu) [reply org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/mapping/MDMDenormalizedTable.java]: done

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/UpdateReportTypeMapping.java (hunk 3 above)
Comment (hw-zhu) [reply org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/UpdateReportTypeMapping.java]: change to size > 1

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/MDMMetadataSources.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
+package com.amalto.core.storage.hibernate;
```
Comment (lpw82gf) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/MDMMetadataSources.java:1]: No copyright?

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/MDMSchemaCreatorImpl.java [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+        this.schemaFilter = schemaFilter;
+    }
+    
+    public void createFromMetadata(
+            Metadata metadata,
+            ExecutionOptions options,
+            Dialect dialect,
+            Formatter formatter,
+            GenerationTarget... targets) {
+//      super.createFromMetadata(metadata, options, dialect, formatter, targets);
```
Comment (lpw82gf) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/MDMSchemaCreatorImpl.java:75]: Why comment it?

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/MDMXmlMappingBinderAccess.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
+package com.amalto.core.storage.hibernate;
```
Comment (lpw82gf) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/MDMXmlMappingBinderAccess.java:1]: No copyright

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/StorageClassLoaderServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
+package com.amalto.core.storage.hibernate;
```
Comment (lpw82gf) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/StorageClassLoaderServiceImpl.java:1]: No copyright

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/mapping/MDMGroupedSchemaMigratorImpl.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    protected void createTable(
+            Table table,
+            Dialect dialect,
+            Metadata metadata,
+            Formatter formatter,
+            ExecutionOptions options,
+            GenerationTarget... targets) {
+        applySqlStrings(
+                false,
+//                dialect.getTableExporter().getSqlCreateStrings( table, metadata ),
```
Comment (lpw82gf) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/mapping/MDMGroupedSchemaMigratorImpl.java:126]: If not needed, just remove

**Review by lpw82gf — APPROVED** [review]

**Pull request merged**